### PR TITLE
Add file changes feature with glob support

### LINKED ISSUE #72748: Add "changes" to command and shell modules
##### SUMMARY
Currently the command and shell modules contain parameters for "creates" and "removes", but not for "changes". The last one could be used to make ansible aware of a native command updating a file and correctly representing the task as changed.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
lib/ansible/modules/shell.py
lib/ansible/modules/command.py

##### ADDITIONAL INFORMATION

With this additional parameter the task changed output can be utilized much better to reflect actual changes and not report "false positives".
This will also enable notify to be used with command and shell modules.

Example usage:
```
- shell: gawk -i inplace '$5 >= 3071' /etc/ssh/moduli
  register: moduli
  changes: /etc/ssh/moduli
  notify:
   - restart sshd
```

### PR DESCRIPTION
**Pull request by imjoseangel**
##### SUMMARY

Fixes #72748

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

command.py

##### ADDITIONAL INFORMATION

Add 'changes' option under shell and command modules with tests

```json
changed: [localhost] => {
    "changed": true,
    "cmd": "perl -pi -e 's/win/won/g' afile.txt",
    "delta": "0:00:00.009514",
    "end": "2021-01-01 16:53:22.693535",
    "invocation": {
        "module_args": {
            "_raw_params": "perl -pi -e 's/win/won/g' /Users/imjoseangel/source/sandbox/afile.txt",
            "_uses_shell": true,
            "argv": null,
            "changes": "/Users/imjoseangel/source/sandbox/afile.*",
            "chdir": "/Users/imjoseangel/source/sandbox",
            "creates": null,
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": false
        }
    },
    "rc": 0,
    "start": "2021-01-01 16:53:22.684021",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "",
    "stdout_lines": []
}
```

### CODE REVIEW

File: lib/ansible/modules/command.py [hunk 1]
```diff
@@ -52,6 +52,11 @@
     description:
       - A filename or (since 2.0) glob pattern. If a matching file exists, this step B(will) be run.
     version_added: "0.8"
+  changes:
+    type: path
+    description:
+      - A filename or (since 2.0) glob pattern. If a matching file exists, this step B(will) be run.
```
Comment (samdoran) [inline lib/ansible/modules/command.py]: Since this is a new feature, the "(since 2.0)" statement can be removed. The description is incorrect. This only controls task change reporting, not whether or not the command will be run. The command will always run (except in check mode), but the task will only report changed if the file changed. This is an import difference in behavior from `creates` and `removes` that should be explained.
```suggestion
      - Path to a file or glob pattern. If a match exists and it is a file, the command will B(always) be run except in check mode. If any files matching the path or pattern are changed after running the command, the task will report a change. 
```

File: lib/ansible/modules/command.py [hunk 2]
```diff
@@ -336,14 +344,34 @@ def main():
...
+                changed=False,
+                rc=0
+            )
+        else:
+            basehashes = []
+            for filename in glob.glob(changes):
+                basefile = open(filename, 'rb')
+                basedata = basefile.read(65536)
+                basehash = hashlib.sha1(basedata).hexdigest()
+                basehashes.append(basehash)
```
Comment (samdoran) [inline lib/ansible/modules/command.py]: This should use `AnsibleModule.digest_from_file()` which is more robust since it checks to see if the hashing algorithm is available and fails gracefully if not.

File: lib/ansible/modules/command.py [hunk 3]
```diff
@@ -356,6 +384,17 @@ def main():
         out = out.rstrip(b"\r\n")
         err = err.rstrip(b"\r\n")
 
+    if changes:
+        finalhashes = []
+        for filename in glob.glob(changes):
+            finalfile = open(filename, 'rb')
+            finaldata = finalfile.read(65536)
+            finalhash = hashlib.sha1(finaldata).hexdigest()
+            finalhashes.append(finalhash)
```
Comment (samdoran) [inline lib/ansible/modules/command.py]: Same here: use `AnsibleModule.digest_from_file()`.

File: lib/ansible/modules/command.py [hunk 4]
```diff
@@ -336,14 +344,34 @@ def main():
                 rc=0
             )
 
+    if changes:
+        # do not run the command if the line contains changes=filename
+        # and the filename does not exist.  This allows idempotence
+        # of command executions.
```
Comment (samdoran) [inline lib/ansible/modules/command.py]: This comment can be more succint.
```suggestion
        # Do not run the command and exit now if the file does not exist
```

File: lib/ansible/modules/command.py [hunk 5]
```diff
@@ -86,9 +91,9 @@
     -  If you want to run a command through the shell (say you are using C(<), C(>), C(|), etc), you actually want the M(ansible.builtin.shell) module instead.
        Parsing shell metacharacters can lead to unexpected commands being executed if quoting is not done correctly so it is more secure to
        use the C(command) module when possible.
-    -  " C(creates), C(removes), and C(chdir) can be specified after the command.
+    -  " C(creates), C(removes), C(changes) and C(chdir) can be specified after the command.
        For instance, if you only want to run a command if a certain file does not exist, use this."
-    -  Check mode is supported when passing C(creates) or C(removes). If running in check mode and either of these are specified, the module will
+    -  Check mode is supported when passing C(creates), C(removes) or C(changes). If running in check mode and either of these are specified, the module will
```
Comment (samdoran) [inline lib/ansible/modules/command.py]: It may be worth clarifying here that `changes=/path/to/file` in check mode will always report `changed=True` even if the command would not result in changes during normal execution. There is no way around this.
```suggestion
    -  Check mode is supported when passing C(creates), C(removes) or C(changes). If running in check mode and any of these are specified, the module will
```

Comment (samdoran) [reply lib/ansible/modules/command.py]: The alternative would be to always report `changed=False` in check mode, but that could result in an unexpected change during normal execution when you were expecting no change. I would rather Ansible report a change in check mode, but do nothing in normal mode to be on the safe side.

File: lib/ansible/modules/shell.py [hunk 6]
```diff
@@ -43,6 +43,11 @@
       - A filename, when it does not exist, this step will B(not) be run.
     type: path
     version_added: "0.8"
+  changes:
+    description:
+      - A filename, when it does not exist, this step will B(not) be run.
```
Comment (samdoran) [inline lib/ansible/modules/shell.py]: See the description comment from the `command` module.

File: lib/ansible/modules/shell.py [hunk 7]
```diff
@@ -77,7 +82,7 @@
     when writing playbooks will follow the trend of using M(ansible.builtin.command)
     unless the M(ansible.builtin.shell) module is explicitly required. When running ad-hoc
     commands, use your best judgement.
-  - Check mode is supported when passing C(creates) or C(removes). If running
+  - Check mode is supported when passing C(creates), C(removes) or C(changes). If running
     in check mode and either of these are specified, the module will check for
```
Comment (samdoran) [inline lib/ansible/modules/shell.py]: See above comments from `command` module.

File: test/integration/targets/command_shell/tasks/main.yml [hunk 8]
```diff
@@ -90,6 +90,18 @@
...
+    src: change_afile.sh
+    dest: "{{ output_dir_test }}"
+    mode: '0755'
+
+- name: prep our test script
+  copy:
+    src: change_test.sh
+    dest: "{{ output_dir_test }}"
+    mode: '0755'
+
```
Comment (samdoran) [inline test/integration/targets/command_shell/tasks/main.yml:104]: These can be combined into a single task using `loop`.

File: test/integration/targets/command_shell/tasks/main.yml [hunk 9]
```diff
@@ -167,6 +179,32 @@
...
+# changes
+
+- name: Verify that afile.txt content has changed
+  assert:
+    that: "'win' in lookup('file', '{{ output_dir_test }}/afile.txt') | b64encode | b64decode"
+
+- name: Changes afile.txt content
+  command: "{{ output_dir_test }}/change_afile.sh {{output_dir_test }}/afile.txt"
+  args:
+    changes: "{{ output_dir_test }}/afile.txt"
```
Comment (samdoran) [inline test/integration/targets/command_shell/tasks/main.yml:191]: The result of this task needs to be registered so that later it can be checked to ensure it reported `change: True`.

File: test/integration/targets/command_shell/tasks/main.yml [hunk 10]
```diff
@@ -167,6 +179,32 @@
     that:
       - command_result3 is not changed
 
+# changes
+
+- name: Verify that afile.txt content has changed
+  assert:
+    that: "'win' in lookup('file', '{{ output_dir_test }}/afile.txt') | b64encode | b64decode"
```
Comment (samdoran) [inline test/integration/targets/command_shell/tasks/main.yml:186]: I'm not sure this task is correct. This isn't really related to the "changes" tests. The "changes" tests  need to operate on an existing file (though a test for `changes` with a file that does not exist would be a good test case to add). This test should create a new file using `copy` with `content` to ensure other tests do not interfere with the "changes" tests.

File: test/integration/targets/command_shell/tasks/main.yml (hunk 9 above)
Comment (samdoran) [reply test/integration/targets/command_shell/tasks/main.yml:191]: Also probably a good idea to use a file just for the "changes" tests rather than reusing `afile.txt`.

File: test/integration/targets/command_shell/tasks/main.yml [hunk 11]
```diff
@@ -167,6 +179,32 @@
...
+
+- name: Changes afile.txt content
+  command: "{{ output_dir_test }}/change_afile.sh {{output_dir_test }}/afile.txt"
+  args:
+    changes: "{{ output_dir_test }}/afile.txt"
+
+- name: Verify that afile.txt content has changed
+  assert:
+    that: "'won' in lookup('file', '{{ output_dir_test }}/afile.txt') | b64encode | b64decode"
```
Comment (samdoran) [inline test/integration/targets/command_shell/tasks/main.yml:195]: A minor nit: group all the `assert` cases at the end of the test rather than interspersing them.

File: test/integration/targets/command_shell/tasks/main.yml [hunk 12]
```diff
@@ -167,6 +179,32 @@
...
+- name: re-run previous command using creates with globbing
+  command: "{{ output_dir_test }}/change_afile.sh {{output_dir_test }}/afile.txt"
+  args:
+    changes: "{{ output_dir_test }}/afile.*"
+  register: command_result4
+
+- name: assert that changes with globbing is working
+  assert:
+    that:
+      - command_result4 is not changed
```
Comment (samdoran) [inline test/integration/targets/command_shell/tasks/main.yml:206]: We also need tests cases for check mode with `changes`.

File: test/integration/targets/command_shell/tasks/main.yml [hunk 13]
```diff
@@ -282,6 +320,22 @@
     path: "{{ output_dir_test }}/afile.txt"
     state: file
 
+# changes
+
+- name: Verify that afile.txt content has changed
+  assert:
+    that: "'win' in lookup('file', '{{ output_dir_test }}/afile.txt') | b64encode | b64decode"
```
Comment (samdoran) [inline test/integration/targets/command_shell/tasks/main.yml:327]: Same comment as above: verifying before the tasks run seem to be verifying the result of previous tests, not tests related to `changes`.

File: test/integration/targets/command_shell/tasks/main.yml [hunk 14]
```diff
@@ -282,6 +320,22 @@
...
+- name: Execute the test.sh script with chdir
+  shell: "{{ output_dir_test }}/change_test.sh > {{ output_dir_test }}/afile.txt"
+  args:
+    chdir: "{{ output_dir_test }}"
+    changes: "{{ output_dir_test }}/afile.txt"
+
+- name: Verify that afile.txt content has changed
+  assert:
+    that: "'won' in lookup('file', '{{ output_dir_test }}/afile.txt') | b64encode | b64decode"
+
```
Comment (samdoran) [inline test/integration/targets/command_shell/tasks/main.yml:338]: How come these are at the bottom? Why not group them with the other changes tests?

Comment (samdoran) [reply test/integration/targets/command_shell/tasks/main.yml:338]: This `main.yml` file is getting a bit long and hard to follow. It's probably a good idea to put these new tests in a separate task file and use `import_tasks` to run them. The rest of the test cases should be broken into separate tasks files, too, but not in this PR.

**Review by samdoran — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/command.py (hunk 2 above)
Comment (imjoseangel) [reply lib/ansible/modules/command.py]: Cool! Thanks for the suggestion. Changed

File: lib/ansible/modules/command.py (hunk 3 above)
Comment (imjoseangel) [reply lib/ansible/modules/command.py]: Cool! Thanks for the suggestion. Changed

File: lib/ansible/modules/command.py (hunk 5 above)
Comment (imjoseangel) [reply lib/ansible/modules/command.py]: We need to double check this as the module always returns `False` as it never knows if the file is going to change. Do you think it is better to remove from `check mode`?

Let me see if there is a way to return a real status on `check mode`

File: lib/ansible/modules/shell.py (hunk 7 above)
Comment (imjoseangel) [reply lib/ansible/modules/shell.py]: We need to check from above if it is a good idea having the `changes` option under `check mode`. The more I think in removing it from `check mode` the more I like the idea.

File: test/integration/targets/command_shell/tasks/main.yml (hunk 14 above)
Comment (imjoseangel) [reply test/integration/targets/command_shell/tasks/main.yml:338]: @samdoran I'm realizing that we have this amount of test in some of the modules already in place. Could be a good idea to work on it and split them in different files per functionality using different PRs? Of course, I will do for this one, but I had the same comment with you for `lineinfile` -> https://github.com/ansible/ansible/pull/70647

What do you think?

**Comment by bcoca** [comment]
note that in the past we've rejected changes that push to modules the functionality of `changed_when`, even if we were to revise that I would not create a specific 'changes' that only deals with files but would have to be a generic ruleset and apply a standard most modules would use.

**Comment by imjoseangel** [comment]
> note that in the past we've rejected changes that push to modules the functionality of `changed_when`, even if we were to revise that I would not create a specific 'changes' that only deals with files but would have to be a generic ruleset and apply a standard most modules would use.

Hi @bcoca I don't catch your comment. Is there anything I need to take into account regarding current changes? Could you please refer to the code if so?

Thank you!

**Comment by samdoran** [comment]
After going through the review of this and thinking through it more, I'm inclined to reject this feature. Not because of any problems in this PR, but due to inherent problems in feature.

With `creates` and `removes`, the command is prevented from running unnecessarily. With `changes`, the command will always be run.

The main point of these parameters is to _prevent_ the command from running. Since it's not possible to know if an arbitrary command would change a file without running the command, I don't think this feature adds enough value to justify including it.

Sorry it took working through this to come to that realization. I'll get some other folks to weigh in and see what they think.

**Comment by imjoseangel** [comment]
No problem from my side. I always learn a lot from you guys.

I'm with you to have a different approach like running the command and check if a file was created in a separate task in a playbook.

**Comment by nitzmahone** [comment]
Yeah, this really seems like a fancified `changed_when` that would only invite a lot more asks for different behavior (you've already shown two- mtime vs hash). If you really need something like this, it's pretty trivial to write a wrapper action that calls arbitrary module(s) and applies whatever logic you want to calculate `changed` on the final result.

**Comment by imjoseangel** [comment]
@samdoran  Probably is good closing #72748?
